examples/static_map/custom_type.cu

```diff
@@ -0,0 +1,128 @@
+/*
+ * Copyright (c) 2021, NVIDIA CORPORATION.
+ *
+ * Licensed under the Apache License, Version 2.0 (the "License");
+ * you may not use this file except in compliance with the License.
+ * You may obtain a copy of the License at
+ *
+ *     http://www.apache.org/licenses/LICENSE-2.0
+ *
+ * Unless required by applicable law or agreed to in writing, software
+ * distributed under the License is distributed on an "AS IS" BASIS,
+ * WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+ * See the License for the specific language governing permissions and
+ * limitations under the License.
+ */
+
+#include <thrust/device_vector.h>
+#include <thrust/logical.h>
+#include <thrust/transform.h>
+
+#include <cuco/static_map.cuh>
+
+// User-defined key type
+#ifdef BUILD_PASCAL_CODE
+// Manual alignment required due to WAR libcu++ bug where cuda::atomic fails for underaligned types
+struct alignas(8) custom_key_type {
+  int32_t a;
+  int32_t b;
+
+  __host__ __device__ custom_key_type() {}
+  __host__ __device__ custom_key_type(int32_t x) : a{x}, b{x} {}
+
+  // Device equality operator is mandatory
+  __device__ bool operator==(custom_key_type const& other) const
+  {
+    return a == other.a and b == other.b;
+  }
```

Comment:
I think it would be more interesting for this example to pass a custom binary operator for the equality comparison as opposed to using `operator==` through the `thrust::equal_to` default.

Reply:
A custom equality operator is indeed used in this example. During insertion, `compare_exchange_strong` will implicitly invoke this `==` operator to check two keys for equality thus it's mandatory.

Reply:
https://godbolt.org/z/M1756bjYo

`compare_exchange_strong` doesn't use `operator==`. It uses bitwise equality comparison (which is why we have that `bitwise_equals` function).

Reply:
https://godbolt.org/z/fqjT31cxj
nvcc 11.5 will work just fine but not the versions before. Custom type with 1 single member variable will work but it's not the case for 2 or more member variables.

Reply:
If you use the latest libcu++ with 11.4 it will work: https://godbolt.org/z/v64WxrP7v

Reply:
https://godbolt.org/z/b9vhhPhf7 data type matters also.

Reply:
Hm, this is a libcu++ bug. See https://github.com/NVIDIA/libcudacxx/issues/223